content/posts/2020/02/thai-in-2562/index.eng.md

```diff
@@ -0,0 +1,93 @@
+
+As such, the world uses various calendars depending on the cultural region. So when displaying user interface, time should be displayed in the appropriate date format for each region. In fact, `DateTimeOffset.ToString()` method has an overload that also receives [`IFormatProvider`][IFormatProvider] objects as parameters for this purpose. [`CultureInfo`][CultureInfo] is the most common class to implement the `IFormatProvider` interface. As the name suggests, `CultureInfo` is the same concept that the Unix family calls locale. As shown below, the result of `DateTimeOffset.ToString()` method depends on which locale you set up as the parameter.
+
+```csharp
+> using System.Globalization;
+> var now = DateTimeOffset.Now;
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ", new CultureInfo("ko-KR"))
+"2020-02-13T17:37:16.436163Z"
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ", new CultureInfo("th-TH"))
+"2563-02-13T17:37:16.436163Z"
+```
+
+However, if you omit the parameter without setting any locale, the method will follow the locale of the environment in which the code is executed. The code below is the result of setting up the operating system region in Korea.
+
+```csharp
+> now.ToString("yyyy-MM-ddTHH:mm:ss.ffffffZ")
+"2020-02-13T17:37:16.436163Z"
+```
+
+[According to the docs, the overload with the omitted parameter][2] follows [`CultureInfo.CurrentCulture`][CultureInfo.CurrentCulture]. As you can infer from its name, the `CultureInfo.CurrentCulture` property points to the locale of the execution environment. Therefore, you must explicitly specify [`CultureInfo.InvariantCulture`][CultureInfo.InvariantCulture] if you want a deterministic action at all times regardless of the locale of your execution environment.
+
+Even though the method may be non-deterministic, the API designed to follow the locale of the execution environment is probably intended because such formatting operations are usually used for user interface, and coding can naturally look appropriate to the cultural community without much concern about internationalization. But the reason why we used this method was not for user interface, but for the cryptographic hash input that had to be deterministic— this turned out to be a mistake.
+
+Now that we know the cause, we have solved the urgent problem by [finding a method that has the `CultureInfo` or `IFormatProvider` parameters omitted, and patching it to explicitly designate `CultureInfo.InvariantCulture`, just like the method `DateTimeOffset.ToString`()][libplanet#734].
+
+CI has also been reinforced with unit testing in Arabic, French, Hebrew locale and so on. Since there are a lot of countries in Europe that use comma (`,`) instead of period (`.`) in decimal places, and countries in the Middle East that writes from right to left, we deliberately chose language regions that were somewhat unfamiliar to us.
```

Comment:
```suggestion
CI has also been reinforced with unit testing in Arabic, French, Hebrew locale and so on. Since there are a lot of countries in Europe that use comma (`,`) instead of a period (`.`) in decimal places, and countries in the Middle East that writes from right to left, we deliberately chose language regions that were somewhat unfamiliar to us.
```